app/graphql/mutations/add_tag.rb

```diff
@@ -0,0 +1,17 @@
+module Mutations
+  class AddTag < BaseMutation
+    # arguments passed to the `resolved` method
+    argument :camp_id, Integer, required: true
+    argument :tag, String, required: true
+
+    type [Types::TagType]
+
+    def resolve(camp_id: nil, tag: nil)
+      camp = Camp.find(camp_id.to_i)
```

Comment:
If it's going to be a common pattern of supporting GraphQL actions as well as Rest actions, we should start considering breaking out the actions into services, since we'll be duplicating all of our business logic this way. (this is a duplicate of the behaviour camps_controller#tag)

Something like this might feel better:
```
# camps_service.rb
def self.tag(camp:, params:)
  ... # add tags
end
```
```
# camps_controller
def tag
  CampService.tag(camp: @camp, params: tag_params)
  render json: @camp.reload.tags
end
```
```
# mutations/add_tag.rb
def resolve(camp_id:, tag:)
  CampService.tag(camp: Camp.find(camp_id), params: { tags: tag })
end
```

Reply:
The GraphQL API is mostly intended to integrate with Realities and a potential mobile app which are Appollo/React based. There is no intention to replicate all controller functionality in graphql queries and mutations as far as I am aware. Where would the CampService live?

Reply:
Loomio's pattern would be to have an `app/services/camp_service.rb`, which handles all requests related to mutating a Camp (usually `create`, `update`, `destroy`, plus any custom actions like `tag`)

I've also seen a UseCase pattern, which will have individual classes which perform a single action, like
```
# app/use_cases/tag_camp.rb
class TagCamp
  def perform(camp:, tags:)
    # add a tag to a camp
  end
end
```
```
TagCamp.perform(camp: @camp, tags: tag_params)
```

Still other places I've worked have mixed in functionality like this via modules
```
CampsController
  include TagCamp
  
  def tag
    tag_camp(@camp, tags_params
    render json: ...
  end
end
```
```
module TagCamp
  def tag_camp(camp, tags)
    ...
  end
end
```

In any case, this suggestion is basically "make sure that we're not performing the same action in two different places, because if if we change how that action works it'll be much harder to get it right"

Reply:
I understand and agree with the overall thought. In this case it is literally a single method call to a library that gets shared. Will definitely reconsider if the pattern continues.